ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -298,18 +306,30 @@ void fillChunks() {
             channelReadBuffer = chunkFillerChannel.getNextChunk(0);
           }
           if (channelReadBuffer != null) {
+            maybeStopTrackingWaitForChannelDataTime();
             chunkToFill = getChunkToFill();
             if (chunkToFill == null) {
-              // no chunks are free to be filled yet.
+              // channel has data, but no chunks are free to be filled yet.
+              maybeStartTrackingWaitForChunkTime();
               break;
-            }
-            bytesFilledSoFar += chunkToFill.fillFrom(channelReadBuffer);
-            if (!channelReadBuffer.hasRemaining()) {
-              chunkFillerChannel.resolveOldestChunk(null);
-              channelReadBuffer = null;
+            } else {
+              // channel has data, and there is a chunk that can be filled.
+              maybeStopTrackingWaitForChunkTime();
+              bytesFilledSoFar += chunkToFill.fillFrom(channelReadBuffer);
+              if (chunkToFill.isReady()) {
+                updateChunkFillerWaitTimeMetrics();
+              }
+              if (!channelReadBuffer.hasRemaining()) {
+                chunkFillerChannel.resolveOldestChunk(null);
+                channelReadBuffer = null;
+              }
             }
           } else {
             // channel does not have more data yet.
+            if (getFreeChunk() != null) {
+              // this means there is a chunk available to be filled, but no data in the channel.
+              maybeStartTrackingWaitForChannelDataTime();
```

Comment:
Would this time also includes the time spent on filling chunks of other `PutOperations`? For example, after `maybeStartTrackingWaitForChannelDataTime()` is called, the chunkFiller thread will leave this `PutOperation` (say PutOperation A) and process other `PutOperations`. In case the number of `PutOperations` is large and chunk size is set large, the time to get back to PutOperation A could be non-negligible but not very relevant to this metric.